ui next: Nodes Comparison Table

mrtracy opened the pull request:
Implement the Nodes Comparison table in UI next.  The implementation is
considerably different, and hopefully improved, when compared to the reference
mithril implementation.

Commit primarily creates a `SortedTable` component which displays tabular data,
and fleshes out the `NodesMain` component to display node status data in the
table. Includes tests for SortedTable component.

Most existing "table" components in react are extremely simple or somewhat
specialized; it was quite easy to create a custom component which meets our
exact requirements.

Facebook's own "fixed-data-table" was briefly considered, but it is a
complicated component and, more importantly, its output does not match our
current CSS. However, we did take cues from the design of "fixed-data-table";
primarily, the table component has no knowledge of the data type being displayed
in its rows. Instead, it simply accepts a count of rows; the "cell" callback for
each column is called with an integer, indicating a row number.

`SortableTable` has a header row, and each column header can be clicked to indicate
that the table data should be sorted in terms of that column. However, the
`SortedTable` component does not sort the data itself - it simply indicates to
higher-level components that the user has requested the data to be sorted for
display.

The table also has a secondary "rollup" header: columns can use it to display a
"rollup" value, which is typically a total or average over all values.
`SortedTable` is not responsible for computing these values: it is up to a
higher level component to compute and provide rollups to the table component.

The Nodes page now contains a `SortedTable`, which displays node status data.
To sort the data and compute rollups, the nodes page makes use of several
Reselect selectors; the data is thus efficiently computed based on the state of
the redux store. This is the same concept as the Computed properties used by our
current admin UI.

The included tests for `SortedTable` make use of the excellent `enyzme` library
to verify the output and behavior of react components.

The following third party libraries have been added:
- Moment.js
- D3.js
- Reselect
- enzyme (test only)
- Sinon.js (test only)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6308)

<!-- Reviewable:end -->

Commit:
ui next: Nodes Comparison Table

Implement the Nodes Comparison table in UI next.  The implementation is
considerably different, and hopefully improved, when compared to the reference
mithril implementation.

Commit primarily creates a `SortedTable` component which displays tabular data,
and fleshes out the `NodesMain` component to display node status data in the
table. Includes tests for SortedTable component.

Most existing "table" components in react are extremely simple or somewhat
specialized; it was quite easy to create a custom component which meets our
exact requirements.

Facebook's own "fixed-data-table" was briefly considered, but it is a
complicated component and, more importantly, its output does not match our
current CSS. However, we did take cues from the design of "fixed-data-table";
primarily, the table component has no knowledge of the data type being displayed
in its rows. Instead, it simply accepts a count of rows; the "cell" callback for
each column is called with an integer, indicating a row number.

`SortableTable` has a header row, and each column header can be clicked to indicate
that the table data should be sorted in terms of that column. However, the
`SortedTable` component does not sort the data itself - it simply indicates to
higher-level components that the user has requested the data to be sorted for
display.

The table also has a secondary "rollup" header: columns can use it to display a
"rollup" value, which is typically a total or average over all values.
`SortedTable` is not responsible for computing these values: it is up to a
higher level component to compute and provide rollups to the table component.

The Nodes page now contains a `SortedTable`, which displays node status data.
To sort the data and compute rollups, the nodes page makes use of several
Reselect selectors; the data is thus efficiently computed based on the state of
the redux store. This is the same concept as the Computed properties used by our
current admin UI.

The included tests for `SortedTable` make use of the excellent `enyzme` library
to verify the output and behavior of react components.

The following third party libraries have been added:
+ Moment.js
+ D3.js
+ Reselect
+ enzyme (test only)
+ Sinon.js (test only)

tamird commented on ui/next/Makefile:
```diff
@@ -54,4 +54,4 @@ jspm.installed: npm.installed
 test:
 # Execute in /app directory, so that ts-node can pick up tsconfig.json.
-	cd app && mocha **/*.spec.ts --require ts-node/register
+	cd app && mocha "**/*.spec.ts*" --require ts-node/register
```
i think you want single quotes here

tamird commented on ui/next/app/util/convert.ts:
```diff
@@ -0,0 +1,7 @@
+/**
+ * NanosToMilli converts a nanoseconds to milliseconds, for use in Java
```
this comment has problems

maxlang commented:
It's a shame Facebook's fixed-data-table didn't work out. Might be worth considering again, especially if we start showing results. I'd rather we don't implement lazy row rendering ourselves.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some minor comments

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[ui/next/app/components/sortabletable.spec.tsx, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI2v3b7HrggLP9oSGm:-KGI2v3b7HrggLP9oSGn:795073444) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.spec.tsx#L15)):*
indentation

---

*[ui/next/app/components/sortabletable.spec.tsx, line 20 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI2woh0RQv1d1umfdy:-KGI2woh0RQv1d1umfdz:795073444) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.spec.tsx#L20)):*
indentation

---

*[ui/next/app/components/sortabletable.spec.tsx, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI3YmvJ-xb9I58Srmp:-KGI3YmvJ-xb9I58Srmq:2091377100) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.spec.tsx#L44)):*
Eventually we won't want the rollup header to appear if no column has a rollup, but this is fine for now.

---

*[ui/next/app/components/sortabletable.tsx, line 22 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHyZeG795g8IpUIHAr:-KGHyZeHnDUTieAHSg-a:1251496972) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L22)):*
I assume you removed section because it isn't in use.

---

*[ui/next/app/components/sortabletable.tsx, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHzU2-fc_6J6Vs-KJ3:-KGHzU20-17_kKpXLLBq:1100837251) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L42)):*
Eventually this should probably be `SortSetting[]` so that you can sort by more than one column. Fine for now

---

*[ui/next/app/components/sortabletable.tsx, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHzgV3TAkJJv_BE1DW:-KGHzgV4Vw3giKah2_8m:-1354509093) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L45)):*
Is this the typical pattern for stuff like this?

---

*[ui/next/app/components/sortabletable.tsx, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI0N2V5gs0OGvyDr9x:-KGI0N2WDGgKc-dhwXQd:-1331833413) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L66)):*
How does this not throw a warning? It's not returning `void`. For clarity you could use lodash's `_.identity`

---

*[ui/next/app/components/sortabletable.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI-Xf3H0975-N6r94T:-KGI-Xf3H0975-N6r94U:529513163) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L70)):*
I haven't seen this destructuring assignment before. Cool!

---

*[ui/next/app/components/sortabletable.tsx, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI0mFXMzNWhdXV8o3f:-KGI0mFXMzNWhdXV8o3g:593744665) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L86)):*
Can be simplified to just `ascending`

---

*[ui/next/app/containers/nodes.tsx, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI3tGjchdBpeCHq7uP:-KGI3tGkxb3omjKkDhoO:886339318) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L74)):*
It's definitely convenient that we can pass around `jsx` like we used to pass around `MithrilVirtualElement`s

---

*[ui/next/app/containers/nodes.tsx, line 140 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI5MiWCqIaf4U1dTnK:-KGI5MiXIgyY87JKNb_n:1013168775) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L140)):*
I know we'll use it for nvd3, but it's funny d3 was brought in for this

---

*[ui/next/app/containers/nodes.tsx, line 163 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI5pZkFD2fKszDNyMK:-KGI5pZkFD2fKszDNyML:-1131954566) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L163)):*
Is a string a valid react node?

---

*[ui/next/app/containers/nodes.tsx, line 203 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI6o7nXlplUeqfIx_K:-KGI6o7oxyvORlCy0k0_:-820807885) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L203)):*
Nice, haven't seen intersection types before.

---

*[ui/next/app/containers/nodes.tsx, line 218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI7PZ_bxRD7OP2arZZ:-KGI7PZa3AuYO-fj0v7o:1559654199) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L218)):*
So if we wanted to allow people to add/remove columns, we'd make `columnDescriptors` a selector and then compose it as well?

---

*[ui/next/app/containers/nodes.tsx, line 273 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI9VrrZRWfwH6RHZtR:-KGI9VrrZRWfwH6RHZtS:1131378990) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L273)):*
Do these have to be `any`?

---

*[ui/next/app/containers/nodes.tsx, line 276 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGIAAXiGrcJkNaKzHy-:-KGIAAXiGrcJkNaKzHy0:990698784) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L276)):*
Could this be simplified to something like `_(columnDescriptors).keyBy("key").mapValues("sort").value();`?

---

*[ui/next/app/redux/nodes.ts, line 14 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHxBB34F2zSvWIvK8-:-KGHxBB34F2zSvWIvK80:-2063815323) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/redux/nodes.ts#L14)):*
Was missing this breaking `fetch`? I'm just confused why you didn't need to import it before (the same thing seemed to have have happened already with `lodash`, but this makes less sense to me.)

---

*[ui/next/app/util/convert.ts, line 2 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGFK38-eK3qNgBJo3fG-r1-2:-KGHxeip_4aB5EN32ORg:1457754971) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/util/convert.ts#L2)):*
It's very special.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-:-KGHyvlEAMu6n5znn2fy:9910032)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 4 of 10 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[ui/next/Makefile, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGFK38-eK3qNgBJo3f9-r1-57:-KGIEPIXGJlF7qixf3ph:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/Makefile#L57)):*
Done.

---

*[ui/next/app/components/sortabletable.spec.tsx, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI2v3b7HrggLP9oSGm:-KGIEaawgGgZwaa6WmYO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.spec.tsx#L15)):*
Done.

---

*[ui/next/app/components/sortabletable.spec.tsx, line 20 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI2woh0RQv1d1umfdy:-KGIEbryfGJUxAGoTHgy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.spec.tsx#L20)):*
Done.

---

*[ui/next/app/components/sortabletable.spec.tsx, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI3YmvJ-xb9I58Srmp:-KGIEeCynvU1Ex6VGL4k:2063968522) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.spec.tsx#L44)):*
Agreed, it's easy logic to add.

---

*[ui/next/app/components/sortabletable.tsx, line 22 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHyZeG795g8IpUIHAr:-KGIEkTMmy1wgjyJGJD1:1799286076) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L22)):*
Yes, we can add it back if it's used again.

---

*[ui/next/app/components/sortabletable.tsx, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHzU2-fc_6J6Vs-KJ3:-KGIEneoiVMphAIoVqX0:-485110604) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L42)):*
Sure.

---

*[ui/next/app/components/sortabletable.tsx, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHzgV3TAkJJv_BE1DW:-KGIEuKXQ3YZS7e25ujO:-1744784198) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L45)):*
Yes, registering callbacks like this is a typical pattern with stateless presentational components.

---

*[ui/next/app/components/sortabletable.tsx, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI0N2V5gs0OGvyDr9x:-KGIGqtdxGR4xGu5JKyG:214702409) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L66)):*
Maybe it's valid to return something from a void function, it's just not valid to use that output.

In either case, I replaced it with (ss) => {}. I wanted the equivalent of python's "pass", which doesn't exist in Javascript, because I was burned out a bit and wasn't thinking straight. An empty function is the correct setting.

---

*[ui/next/app/components/sortabletable.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI-Xf3H0975-N6r94T:-KGIHe006wxP2UoK7ag3:905670926) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L70)):*
Yep, it's neat.

---

*[ui/next/app/components/sortabletable.tsx, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI0mFXMzNWhdXV8o3f:-KGIHoqoSyfZCTGvd27d:-232456216) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/components/sortabletable.tsx#L86)):*
Neat! fixed.

---

*[ui/next/app/containers/nodes.tsx, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI3tGjchdBpeCHq7uP:-KGIHvHwPnJvstF61Ey2:-1099313596) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L74)):*
It's really fantastic.

---

*[ui/next/app/containers/nodes.tsx, line 140 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI5MiWCqIaf4U1dTnK:-KGII02pvEHG44Szt1Cd:548178466) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L140)):*
Agreed, but since we're bringing it in anyway, it makes sense. :)

---

*[ui/next/app/containers/nodes.tsx, line 163 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI5pZkFD2fKszDNyMK:-KGII6jInMXEwvd3aGh5:650200510) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L163)):*
Yes, it's basically the same idea you had with MithrilChild.

---

*[ui/next/app/containers/nodes.tsx, line 218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI7PZ_bxRD7OP2arZZ:-KGIITdQAhwkK-vr0Qk4:-168611395) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L218)):*
Yes, that's one way we could do it.

---

*[ui/next/app/containers/nodes.tsx, line 273 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGI9VrrZRWfwH6RHZtR:-KGIJAYtfyZY7kPhz5IC:-311787824) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L273)):*
They don't HAVE to be, but I showed you my frustration with referencing the AdminUISetting interface - i'll likely take another pass at this later, across all pages, but I'm sick of dealing with it for now.

---

*[ui/next/app/containers/nodes.tsx, line 276 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGIAAXiGrcJkNaKzHy-:-KGIKxJpPo3qizDx2uny:2126261835) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/containers/nodes.tsx#L276)):*
Gave it a shot, "mapValues" isn't available on the chaining wrapper in our typings for some reason. I'm going to leave it as-is.

---

*[ui/next/app/redux/nodes.ts, line 14 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGHxBB34F2zSvWIvK8-:-KGIL1lb-HLtruUK03iZ:-128605932) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/redux/nodes.ts#L14)):*
The typing for "whatwg-fetch" is ambient (it extends Window), but the module still needs to be referenced via an import like this in order to pull in the javascript in the compiled output.

It was working in chrome because chrome supports `fetch` natively.  It wasn't working in other browsers (safari).

---

*[ui/next/app/util/convert.ts, line 2 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308#-KGFK38-eK3qNgBJo3fG-r1-2:-KGILPng2gtIb918s7Xi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6b7b64d3063e0c00906af52183f4eb35cfe3c99/ui/next/app/util/convert.ts#L2)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6308)*
<!-- Sent from Reviewable.io -->
